File: MINV/dminv.f
C   14/11/86 611282107  MEMBER NAME  DMINV    (PETLIB)      FORTRAN77
C
C     ..................................................................
C
C        SUBROUTINE DMINV
C
C        PURPOSE
C           INVERT A MATRIX
C
C        USAGE
C           CALL DMINV(A,N,D,L,M)
C
C        DESCRIPTION OF PARAMETERS
C           A - INPUT MATRIX, DESTROYED IN COMPUTATION AND REPLACED BY
C               RESULTANT INVERSE.
C           N - ORDER OF MATRIX A
C           D - RESULTANT DETERMINANT
C           L - WORK VECTOR OF LENGTH N
C           M - WORK VECTOR OF LENGTH N
C
C        REMARKS
C           MATRIX A MUST BE A GENERAL MATRIX
C
C        SUBROUTINES AND FUNCTION SUBPROGRAMS REQUIRED
C           NONE
C
C        METHOD
C           THE STANDARD GAUSS-JORDAN METHOD IS USED. THE DETERMINANT
C           IS ALSO CALCULATED. A DETERMINANT OF ZERO INDICATES THAT
C           THE MATRIX IS SINGULAR.
C
C        TRANSLATED INTO FORTRAN77 11.86 JKL
C     ..................................................................
C
      SUBROUTINE DMINV(A,N,D,L,M)
      IMPLICIT REAL*8 (A-H,O-Z)
      DIMENSION A(1),L(1),M(1)
      write(*,fmt='(1x, 25d10.3)') (a(i), i=1,N*N)
      D=1.D0
      NK=-N
      DO 80 K=1,N
c write(*, fmt='(4hk   ,i8)') k 
         NK=NK+N
         L(K)=K
         M(K)=K
         KK=NK+K
c write(*, fmt='(4hkk  ,i8)') kk
         BIGA=A(KK)
c      write(*,101) biga, kk
c 101	format(1x,d15.8, i8)
         DO 21 J=K,N
c write(*, fmt='(4hj   ,i8)') j 
            IZ=N*(J-1)
c write(*, fmt='(4hiz  ,i8)') iz
            DO 20 I=K,N
               IJ=IZ+I
c write(*, fmt='(4hij  ,i8)') ij
               IF(DABS(BIGA).LT.DABS(A(IJ))) THEN
                  BIGA=A(IJ)
                  L(K)=I
                  M(K)=J
               ENDIF
   20       CONTINUE
   21    CONTINUE
c       write(*, fmt='(1x, 5i8)') (l(is), is=1,5)
c       write(*, fmt='(1x, 5i8)') (m(is), is=1,5)
         J=L(K)
         IF(J.GT.K) THEN
            KI=K-N
            DO 30 I=1,N
               KI=KI+N
c write(*, fmt='(4hki  ,i8)') ki
               HOLD=-A(KI)
c		       write(*,102) hold
c 102	format(1x,'h1 ', d15.8)
               JI=KI-K+J
c write(*, fmt='(4hji  ,i8)') ji
               A(KI)=A(JI)
               A(JI) =HOLD
   30       CONTINUE
         ENDIF
         I=M(K)
         IF(I.GT.K) THEN
            JP=N*(I-1)
            DO 40 J=1,N
               JK=NK+J
c write(*, fmt='(4hjk  ,i8)') jk
               JI=JP+J
c write(*, fmt='(4hji  ,i8)') ji
               HOLD=-A(JK)
               A(JK)=A(JI)
               A(JI) =HOLD
   40       CONTINUE
         ENDIF
         IF(BIGA.EQ. 0.D0) THEN
            D=0D0
            RETURN
         ENDIF
         DO 55 I=1,N
            IF(I.NE.K) THEN
               IK=NK+I
c write(*, fmt='(4hik  ,i8)') ik
               A(IK)=A(IK)/(-BIGA)
            ENDIF
   55    CONTINUE
         DO 66 I=1,N
            IK=NK+I
c write(*, fmt='(4hik  ,i8)') ik
            HOLD=A(IK)
            IJ=I-N
c write(*, fmt='(4hij  ,i8)') ij
            DO 65 J=1,N
               IJ=IJ+N
c write(*, fmt='(4hij  ,i8)') ij
               IF(I.NE.K.AND.J.NE.K) THEN
                  KJ=IJ-I+K
c write(*, fmt='(4hkj  ,i8)') kj
                  A(IJ)=HOLD*A(KJ)+A(IJ)
               ENDIF
   65       CONTINUE
   66    CONTINUE
         KJ=K-N
c write(*, fmt='(4hkj  ,i8)') kj
         DO 75 J=1,N
            KJ=KJ+N
c write(*, fmt='(4hkj  ,i8)') kj
            IF(J.NE.K) A(KJ)=A(KJ)/BIGA
   75    CONTINUE
         D=D*BIGA
c write(*, fmt='(4hkk  ,i8)') kk
         A(KK)=1D0/BIGA
   80 CONTINUE
      DO 90 K=N-1,1,-1
         I=L(K)
         IF(I.GT.K) THEN
            JQ=N*(K-1)
            JR=N*(I-1)
            DO 110 J=1,N
               JK=JQ+J
c write(*, fmt='(4hjk  ,i8)') jk
               HOLD=A(JK)
               JI=JR+J
c write(*, fmt='(4hji  ,i8)') ji
               A(JK)=-A(JI)
               A(JI) =HOLD
  110       CONTINUE
         ENDIF
         J=M(K)
         IF(J.GT.K) THEN
            KI=K-N
            DO 130 I=1,N
               KI=KI+N
c write(*, fmt='(4hki  ,i8)') ki
               HOLD=A(KI)
               JI=KI-K+J
c write(*, fmt='(4hji  ,i8)') ji
               A(KI)=-A(JI)
               A(JI) =HOLD
  130       CONTINUE
         ENDIF
   90 CONTINUE
      RETURN
      END

File: MINV/dminv.cxx
//   14/11/86 611282107  member name  dminv    (petlib)      fortran77
//
//     .................................................................
//
//        subroutine dminv
//
//        purpose
//           invert a matrix
//
//        usage
//           double d = dminv(a,n)
//
//        description of parameters
//           a - input matrix, destroyed in computation and replaced by
//               resultant inverse.
//           n - order of matrix a
//           d - resultant determinant
//           l - work vector of length n
//           m - work vector of length n
//
//        remarks
//           matrix a must be a general matrix
//
//        subroutines and function subprograms required
//           none
//
//        method
//           the standard gauss-jordan method is used. the determinant
//           is also calculated. a determinant of zero indicates that
//           the matrix is singular.
//
//        translated into fortran77 11.86 jkl
//        translated into c++       05.98 jkl
//     .................................................................
//

#include <math.h>
#include <stdio.h>


double dminv(double *a, int n)
{
      int * l = new int[n];
      int * m = new int[n];
      double d=1.0;
	printf("matrix of %d\n",n);
	for(int i=0; i<n*n; i++) printf("%3.1lf  ", a[i]);
	printf("\n");

      int nk=-n;
      for(int k=0; k<n; k++)
      {
	 printf("k  %8d\n", k +1);
         nk=nk+n;
         l[k]=k;
         m[k]=k;
         int kk=nk+k;
	 printf("kk %8d\n", kk+1);
         double biga=a[kk];
// 	 printf("%15.8e %d\n", biga, kk);
	 for(int j=k; j<n; j++)
	 {
	 printf("j  %8d\n", j +1);
            int iz=n*j;
	 printf("iz %8d\n", iz);
	    for(int i=k; i<n; i++)
	    {
               int ij=iz+i;
	 printf("ij %8d\n", ij+1);
               if(fabs(biga) < fabs(a[ij]))
	       {
                  biga=a[ij];
                  l[k]=i;
                  m[k]=j;
               }
	    }
	 }
	 for(int s=0; s<5; s++) printf("%8d  ", l[s]+1); printf("\n");
	 for(int s=0; s<5; s++) printf("%8d  ", m[s]+1); printf("\n");
         int j=l[k];
         if(j > k) {
            int ki=k-n;
	    for(int i=0; i<n; i++)
	    {
               ki=ki+n;
	 printf("ki %8d\n", ki+1);
               double hold = -a[ki];
// 		  printf("h1 %15.8e\n", hold);
               int ji=ki-k+j;
	 printf("ji %8d\n", ji+1);
               a[ki]=a[ji];
               a[ji] =hold;
	    }
         }
         int i=m[k];
         if(i > k) {
            int jp=n*i;
	    for(int j=0; j<n; j++)
	    {
               int jk=nk+j;
	 printf("jk %8d\n", jk+1);
               int ji=jp+j;
	 printf("ji %8d\n", ji+1);
               double hold= - a[jk];
               a[jk]=a[ji];
               a[ji] =hold;
	    }
         }
         if(biga == 0.0) {
	    delete [] l;
	    delete [] m;
            return 0.0;
         }
	 for(int i=0; i<n; i++)
	 {
            if(i != k) {
               int ik=nk+i;
	 printf("ik %8d\n", ik+1);
               a[ik]=a[ik]/(-biga);
            }
	 }

	 for(i=0; i<n; i++)
	 {
            int ik=nk+i;
	 printf("ik %8d\n", ik+1);
            double hold=a[ik];
            int ij=i-n;
	 printf("ij %8d\n", ij+1);
	    for(int j=0; j<n; j++)
	    {
               ij=ij+n;
	 printf("ij %8d\n", ij+1);
               if(i != k && j != k) {
                  int kj=ij-i+k;
	 printf("kj %8d\n", kj+1);
                  a[ij]=hold*a[kj]+a[ij];
               }
	    }
	 }
         int kj=k-n;
	 printf("kj %8d\n", kj+1);
	 for(j=0; j<n; j++)
	 {
            kj=kj+n;
	 printf("kj %8d\n", kj+1);
            if(j != k) a[kj]=a[kj]/biga;
	 }
         d=d*biga;
	 printf("kk %8d\n", kk+1);
         a[kk]=1.0/biga;
      }

      for(int k=n-2; k>=0; k--)
      {
         int i=l[k];
         if(i > k) {
            int jq=n*k;
            int jr=n*i;
	    for(int j=0; j<n; j++)
	    {
               int jk=jq+j;
	 printf("jk %8d\n", jk+1);
               double hold=a[jk];
               int ji=jr+j;
	 printf("ji %8d\n", ji+1);
               a[jk]=-a[ji];
               a[ji] =hold;
	    }
         }
         int j=m[k];
         if(j > k) {
            int ki=k-n;
	    for(int i=0; i<n; i++)
	    {
               ki=ki+n;
	 printf("ki %8d\n", ki+1);
               double hold=a[ki];
               int ji=ki-k+j;
	 printf("ji %8d\n", ji+1);
               a[ki]=-a[ji];
               a[ji] =hold;
	    }
         }
      }
      delete [] l;
      delete [] m;
      return d;
}
